File: __tests__/expect_test.ml
open Jest

type test_record = {
  value: string
}

let () =

describe "Expect" (fun () -> 
  let open Expect in

  test "toBe" (fun () ->
    expect (1 + 2) |> toBe 3);
  test "toBeCloseTo" (fun () ->
    expect (1. +. 2.) |> toBeCloseTo 3.);
  test "toBeSoCloseTo" (fun () ->
    expect (1. +. 2.123) |> toBeSoCloseTo 3.123 ~digits:3);
  test "toBeGreaterThan" (fun () ->
    expect 4 |> toBeGreaterThan 3);
  test "toBeGreaterThanOrEqual" (fun () ->
    expect 4 |> toBeGreaterThanOrEqual 4);
  test "toBeLessThan" (fun () ->
    expect 4 |> toBeLessThan 5);
  test "toBeLessThanOrEqual" (fun () ->
    expect 4 |> toBeLessThanOrEqual 4);
  test "toBeSuperSetOf" (fun () ->
    expect [| "a"; "b"; "c" |] |> toBeSupersetOf [| "a"; "c" |]);
  test "toContain" (fun () ->
    expect [| "a"; "b"; "c" |] |> toContain "b");
  test "toContainEqual" (fun () ->
    expect [| {value = "a"}; {value = "b"}; {value = "c"} |] |> toContainEqual {value = "b"});
  test "toContainString" (fun () ->
    expect "banana" |> toContainString "nana");
  test "toHaveLength" (fun () ->
    expect [| "a"; "b"; "c" |] |> toHaveLength 3);
  test "toEqual" (fun () ->
    expect (1 + 2) |> toEqual 3);
  test "toMatch" (fun () ->
    expect "banana" |> toMatch "nana");
  test "toMatchRe" (fun () ->
    expect "banana" |> toMatchRe [%re "/ana/"]);
  test "toThrow" (fun () ->
    expect (fun () -> assert false) |> toThrow);

  test "toMatchInlineSnapshot" (fun () ->
    expect "foo" |> toMatchInlineSnapshot "\"foo\"");
  test "toMatchSnapshot" (fun () ->
    expect "foo" |> toMatchSnapshot);
  test "toMatchSnapshotWithName" (fun () ->
    expect "foo" |> toMatchSnapshotWithName "bar");
  test "toThrowErrorMatchingSnapshot" (fun () ->
    expect (fun () -> Js.Exn.raiseError "foo error") |> toThrowErrorMatchingSnapshot);

  test "not toBe" (fun () ->
    expect (1 + 2) |> not_ |> toBe 4);
  test "not toBeCloseTo" (fun () ->
    expect (1. +. 2.) |> not_ |> toBeCloseTo 4.);
  test "not toBeSoCloseTo" (fun () ->
    expect (1. +. 2.123) |> not_ |> toBeSoCloseTo 3.12 ~digits:3);
  test "not toBeGreaterThan" (fun () ->
    expect 4 |> not_ |> toBeGreaterThan 4);
  test "not toBeGreaterThanOrEqual" (fun () ->
    expect 4 |> not_ |> toBeGreaterThanOrEqual 5);
  test "not toBeLessThan" (fun () ->
    expect 4 |> not_ |> toBeLessThan 4);
  test "not toBeLessThanOrEqual" (fun () ->
    expect 4 |> not_ |> toBeLessThanOrEqual 3);
  test "not toBeSuperSetOf" (fun () ->
    expect [| "a"; "b"; "c" |] |> not_ |> toBeSupersetOf [| "a"; "d" |]);
  test "not toContain" (fun () ->
    expect [| "a"; "b"; "c" |] |> not_ |> toContain "d");
  test "not toContainEqual" (fun () ->
    expect [| {value = "a"}; {value = "b"}; {value = "c"} |] |> not_ |> toContainEqual {value = "d"});
  test "not toContainString" (fun () ->
    expect "banana" |> not_ |> toContainString "nanan");
  test "not toHaveLength" (fun () ->
    expect [| "a"; "b"; "c" |] |> not_ |> toHaveLength 2);
  test "not toEqual" (fun () ->
    expect (1 + 2) |> not_ |> toEqual 4);
  test "not toMatch" (fun () ->
    expect "banana" |> not_ |> toMatch "nanan");
  test "not toMatchRe" (fun () ->
    expect "banana" |> not_ |> toMatchRe [%re "/anas/"]);
  test "not toThrow" (fun () ->
    expect (fun () -> 2) |> not_ |> toThrow);

  test "expectFn" (fun () ->
    expectFn raise (Invalid_argument "foo") |> toThrow);
);

describe "Expect.Operators" (fun () -> 
  let open Expect in
  let open! Expect.Operators in

  test "==" (fun () ->
    expect (1 + 2) == 3);
  test ">" (fun () ->
    expect 4 > 3);
  test ">=" (fun () ->
    expect 4 >= 4);
  test "<" (fun () ->
    expect 4 < 5);
  test "<=" (fun () ->
    expect 4 <= 4);
  test "=" (fun () ->
    expect (1 + 2) = 3);
  test "<>" (fun () ->
    expect (1 + 2) <> 4);
  test "!=" (fun () ->
    expect (1 + 2) != 4);
);

describe "ExpectJs" (fun () -> 
  let open ExpectJs in

  test "toBeDefined" (fun () ->
    expect (Js.Undefined.return 3) |> toBeDefined);
  test "toBeFalsy" (fun () ->
    expect nan |> toBeFalsy);
  test "toBeNull" (fun () ->
    expect Js.null |> toBeNull);
  test "toBeTruthy" (fun () ->
    expect [||] |> toBeTruthy);
  test "toBeUndefined" (fun () ->
    expect Js.Undefined.empty |> toBeUndefined);
  test "toContainProperties" (fun () ->
    expect [%obj { foo = 0; bar = true }] |> toContainProperties [| "foo"; "bar" |]);
  test "toMatchObject" (fun () ->
    expect [%obj { a = 1; b = 2; c = 3 }] |> toMatchObject [%obj { a = 1; b = 2 }]);

  test "not toBeDefined" (fun () ->
    expect (Js.undefined) |> not_ |> toBeDefined);
  test "not toBeFalsy" (fun () ->
    expect [||] |> not_ |> toBeFalsy);
  test "not toBeNull" (fun () ->
    expect (Js.Null.return 4) |> not_ |> toBeNull);
  test "not toBeTruthy" (fun () ->
    expect nan |> not_ |> toBeTruthy);
  test "not toBeUndefined" (fun () ->
    expect (Js.Undefined.return 4) |> not_ |> toBeUndefined);
  test "not toContainProperties" (fun () ->
    expect [%obj { foo = 0; bar = true }] |> not_ |> toContainProperties [| "foo"; "zoo" |]);
  test "not toMatchObject" (fun () ->
    expect [%obj { a = 1; b = 2; c = 3 }] |> not_ |> toMatchObject [%obj { a = 1; c = 2 }]);
);

File: lib/js/__tests__/expect_test.js
'use strict';

var Jest = require("../src/jest.js");
var Curry = require("bs-platform/lib/js/curry.js");
var Js_exn = require("bs-platform/lib/js/js_exn.js");
var Pervasives = require("bs-platform/lib/js/pervasives.js");
var Caml_builtin_exceptions = require("bs-platform/lib/js/caml_builtin_exceptions.js");

describe("Expect", (function () {
        Jest.test("toBe", (function () {
                return Jest.Expect[/* toBe */2](3, Jest.Expect[/* expect */0](3));
              }));
        Jest.test("toBeCloseTo", (function () {
                return Jest.Expect[/* toBeCloseTo */3](3, Jest.Expect[/* expect */0](1 + 2));
              }));
        Jest.test("toBeSoCloseTo", (function () {
                return Jest.Expect[/* toBeSoCloseTo */4](3.123, 3, Jest.Expect[/* expect */0](1 + 2.123));
              }));
        Jest.test("toBeGreaterThan", (function () {
                return Jest.Expect[/* toBeGreaterThan */5](3, Jest.Expect[/* expect */0](4));
              }));
        Jest.test("toBeGreaterThanOrEqual", (function () {
                return Jest.Expect[/* toBeGreaterThanOrEqual */6](4, Jest.Expect[/* expect */0](4));
              }));
        Jest.test("toBeLessThan", (function () {
                return Jest.Expect[/* toBeLessThan */7](5, Jest.Expect[/* expect */0](4));
              }));
        Jest.test("toBeLessThanOrEqual", (function () {
                return Jest.Expect[/* toBeLessThanOrEqual */8](4, Jest.Expect[/* expect */0](4));
              }));
        Jest.test("toBeSuperSetOf", (function () {
                return Jest.Expect[/* toBeSupersetOf */9](/* array */[
                            "a",
                            "c"
                          ], Jest.Expect[/* expect */0](/* array */[
                                "a",
                                "b",
                                "c"
                              ]));
              }));
        Jest.test("toContain", (function () {
                return Jest.Expect[/* toContain */10]("b", Jest.Expect[/* expect */0](/* array */[
                                "a",
                                "b",
                                "c"
                              ]));
              }));
        Jest.test("toContainString", (function () {
                return Jest.Expect[/* toContainString */11]("nana", Jest.Expect[/* expect */0]("banana"));
              }));
        Jest.test("toHaveLength", (function () {
                return Jest.Expect[/* toHaveLength */13](3, Jest.Expect[/* expect */0](/* array */[
                                "a",
                                "b",
                                "c"
                              ]));
              }));
        Jest.test("toEqual", (function () {
                return Jest.Expect[/* toEqual */12](3, Jest.Expect[/* expect */0](3));
              }));
        Jest.test("toMatch", (function () {
                return Jest.Expect[/* toMatch */14]("nana", Jest.Expect[/* expect */0]("banana"));
              }));
        Jest.test("toMatchRe", (function () {
                return Jest.Expect[/* toMatchRe */15]((/ana/), Jest.Expect[/* expect */0]("banana"));
              }));
        Jest.test("toThrow", (function () {
                return Jest.Expect[/* toThrow */18](Jest.Expect[/* expect */0]((function () {
                                  throw [
                                        Caml_builtin_exceptions.assert_failure,
                                        [
                                          "expect_test.ml",
                                          37,
                                          22
                                        ]
                                      ];
                                })));
              }));
        Jest.test("toThrowException", (function () {
                return Jest.Expect[/* toThrowException */20]([
                            Caml_builtin_exceptions.invalid_argument,
                            "foo"
                          ], Jest.Expect[/* expect */0]((function () {
                                  throw [
                                        Caml_builtin_exceptions.invalid_argument,
                                        "foo"
                                      ];
                                })));
              }));
        Jest.test("toThrowMessage", (function () {
                return Jest.Expect[/* toThrowMessage */21]("Invalid_argument,-3,foo", Jest.Expect[/* expect */0]((function () {
                                  throw [
                                        Caml_builtin_exceptions.invalid_argument,
                                        "foo"
                                      ];
                                })));
              }));
        Jest.test("toThrowMessageRe", (function () {
                return Jest.Expect[/* toThrowMessageRe */22]((/Assert_failure/), Jest.Expect[/* expect */0]((function () {
                                  throw [
                                        Caml_builtin_exceptions.assert_failure,
                                        [
                                          "expect_test.ml",
                                          43,
                                          22
                                        ]
                                      ];
                                })));
              }));
        Jest.test("toMatchSnapshot", (function () {
                return Jest.Expect[/* toMatchSnapshot */16](Jest.Expect[/* expect */0]("foo"));
              }));
        Jest.test("toMatchSnapshotWithName", (function () {
                return Jest.Expect[/* toMatchSnapshotWithName */17]("bar", Jest.Expect[/* expect */0]("foo"));
              }));
        Jest.test("toThrowErrorMatchingSnapshot", (function () {
                return Jest.Expect[/* toThrowErrorMatchingSnapshot */19](Jest.Expect[/* expect */0]((function () {
                                  return Js_exn.raiseError("foo error");
                                })));
              }));
        Jest.test("not toBe", (function () {
                return Jest.Expect[/* toBe */2](4, Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0](3)));
              }));
        Jest.test("not toBeCloseTo", (function () {
                return Jest.Expect[/* toBeCloseTo */3](4, Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0](1 + 2)));
              }));
        Jest.test("not toBeSoCloseTo", (function () {
                return Jest.Expect[/* toBeSoCloseTo */4](3.12, 3, Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0](1 + 2.123)));
              }));
        Jest.test("not toBeGreaterThan", (function () {
                return Jest.Expect[/* toBeGreaterThan */5](4, Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0](4)));
              }));
        Jest.test("not toBeGreaterThanOrEqual", (function () {
                return Jest.Expect[/* toBeGreaterThanOrEqual */6](5, Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0](4)));
              }));
        Jest.test("not toBeLessThan", (function () {
                return Jest.Expect[/* toBeLessThan */7](4, Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0](4)));
              }));
        Jest.test("not toBeLessThanOrEqual", (function () {
                return Jest.Expect[/* toBeLessThanOrEqual */8](3, Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0](4)));
              }));
        Jest.test("not toBeSuperSetOf", (function () {
                return Jest.Expect[/* toBeSupersetOf */9](/* array */[
                            "a",
                            "d"
                          ], Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0](/* array */[
                                    "a",
                                    "b",
                                    "c"
                                  ])));
              }));
        Jest.test("not toContain", (function () {
                return Jest.Expect[/* toContain */10]("d", Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0](/* array */[
                                    "a",
                                    "b",
                                    "c"
                                  ])));
              }));
        Jest.test("not toContainString", (function () {
                return Jest.Expect[/* toContainString */11]("nanan", Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0]("banana")));
              }));
        Jest.test("not toHaveLength", (function () {
                return Jest.Expect[/* toHaveLength */13](2, Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0](/* array */[
                                    "a",
                                    "b",
                                    "c"
                                  ])));
              }));
        Jest.test("not toEqual", (function () {
                return Jest.Expect[/* toEqual */12](4, Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0](3)));
              }));
        Jest.test("not toMatch", (function () {
                return Jest.Expect[/* toMatch */14]("nanan", Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0]("banana")));
              }));
        Jest.test("not toMatchRe", (function () {
                return Jest.Expect[/* toMatchRe */15]((/anas/), Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0]("banana")));
              }));
        Jest.test("not toThrow", (function () {
                return Jest.Expect[/* toThrow */18](Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0]((function () {
                                      return 2;
                                    }))));
              }));
        Jest.test("not toThrowException", (function () {
                return Jest.Expect[/* toThrowException */20]([
                            Caml_builtin_exceptions.invalid_argument,
                            "bar"
                          ], Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0]((function () {
                                      throw [
                                            Caml_builtin_exceptions.invalid_argument,
                                            "foo"
                                          ];
                                    }))));
              }));
        Jest.test("not toThrowMessage", (function () {
                return Jest.Expect[/* toThrowMessage */21]("Invalid_argument,-3,foo", Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0]((function () {
                                      throw [
                                            Caml_builtin_exceptions.invalid_argument,
                                            "bar"
                                          ];
                                    }))));
              }));
        Jest.test("not toThrowMessageRe", (function () {
                return Jest.Expect[/* toThrowMessageRe */22]((/Assert_failure!/), Jest.Expect[/* not_ */23](Jest.Expect[/* expect */0]((function () {
                                      throw [
                                            Caml_builtin_exceptions.assert_failure,
                                            [
                                              "expect_test.ml",
                                              87,
                                              22
                                            ]
                                          ];
                                    }))));
              }));
        return Jest.test("expectFn", (function () {
                      return Jest.Expect[/* toThrowException */20]([
                                  Caml_builtin_exceptions.invalid_argument,
                                  "foo"
                                ], Jest.Expect[/* expectFn */1]((function (prim) {
                                        throw prim;
                                      }), [
                                      Caml_builtin_exceptions.invalid_argument,
                                      "foo"
                                    ]));
                    }));
      }));

describe("Expect.Operators", (function () {
        Jest.test("==", (function () {
                return Curry._2(Jest.Expect[/* Operators */24][/* == */0], Jest.Expect[/* expect */0](3), 3);
              }));
        Jest.test(">", (function () {
                return Curry._2(Jest.Expect[/* Operators */24][/* > */1], Jest.Expect[/* expect */0](4), 3);
              }));
        Jest.test(">=", (function () {
                return Curry._2(Jest.Expect[/* Operators */24][/* >= */2], Jest.Expect[/* expect */0](4), 4);
              }));
        Jest.test("<", (function () {
                return Curry._2(Jest.Expect[/* Operators */24][/* < */3], Jest.Expect[/* expect */0](4), 5);
              }));
        Jest.test("<=", (function () {
                return Curry._2(Jest.Expect[/* Operators */24][/* <= */4], Jest.Expect[/* expect */0](4), 4);
              }));
        Jest.test("=", (function () {
                return Curry._2(Jest.Expect[/* Operators */24][/* = */5], Jest.Expect[/* expect */0](3), 3);
              }));
        Jest.test("<>", (function () {
                return Curry._2(Jest.Expect[/* Operators */24][/* <> */6], Jest.Expect[/* expect */0](3), 4);
              }));
        return Jest.test("!=", (function () {
                      return Curry._2(Jest.Expect[/* Operators */24][/* != */7], Jest.Expect[/* expect */0](3), 4);
                    }));
      }));

describe("ExpectJs", (function () {
        Jest.test("toBeDefined", (function () {
                return Jest.ExpectJs[/* toBeDefined */25](Jest.ExpectJs[/* expect */0](3));
              }));
        Jest.test("toBeFalsy", (function () {
                return Jest.ExpectJs[/* toBeFalsy */26](Jest.ExpectJs[/* expect */0](Pervasives.nan));
              }));
        Jest.test("toBeNull", (function () {
                return Jest.ExpectJs[/* toBeNull */27](Jest.ExpectJs[/* expect */0](null));
              }));
        Jest.test("toBeTruthy", (function () {
                return Jest.ExpectJs[/* toBeTruthy */28](Jest.ExpectJs[/* expect */0](/* array */[]));
              }));
        Jest.test("toBeUndefined", (function () {
                return Jest.ExpectJs[/* toBeUndefined */29](Jest.ExpectJs[/* expect */0](undefined));
              }));
        Jest.test("toContainProperties", (function () {
                return Jest.ExpectJs[/* toContainProperties */30](/* array */[
                            "foo",
                            "bar"
                          ], Jest.ExpectJs[/* expect */0]({
                                foo: 0,
                                bar: /* true */1
                              }));
              }));
        Jest.test("toMatchObject", (function () {
                return Jest.ExpectJs[/* toMatchObject */31]({
                            a: 1,
                            b: 2
                          }, Jest.ExpectJs[/* expect */0]({
                                a: 1,
                                b: 2,
                                c: 3
                              }));
              }));
        Jest.test("not toBeDefined", (function () {
                return Jest.ExpectJs[/* toBeDefined */25](Jest.ExpectJs[/* not_ */23](Jest.ExpectJs[/* expect */0](undefined)));
              }));
        Jest.test("not toBeFalsy", (function () {
                return Jest.ExpectJs[/* toBeFalsy */26](Jest.ExpectJs[/* not_ */23](Jest.ExpectJs[/* expect */0](/* array */[])));
              }));
        Jest.test("not toBeNull", (function () {
                return Jest.ExpectJs[/* toBeNull */27](Jest.ExpectJs[/* not_ */23](Jest.ExpectJs[/* expect */0](4)));
              }));
        Jest.test("not toBeTruthy", (function () {
                return Jest.ExpectJs[/* toBeTruthy */28](Jest.ExpectJs[/* not_ */23](Jest.ExpectJs[/* expect */0](Pervasives.nan)));
              }));
        Jest.test("not toBeUndefined", (function () {
                return Jest.ExpectJs[/* toBeUndefined */29](Jest.ExpectJs[/* not_ */23](Jest.ExpectJs[/* expect */0](4)));
              }));
        Jest.test("not toContainProperties", (function () {
                return Jest.ExpectJs[/* toContainProperties */30](/* array */[
                            "foo",
                            "zoo"
                          ], Jest.ExpectJs[/* not_ */23](Jest.ExpectJs[/* expect */0]({
                                    foo: 0,
                                    bar: /* true */1
                                  })));
              }));
        return Jest.test("not toMatchObject", (function () {
                      return Jest.ExpectJs[/* toMatchObject */31]({
                                  a: 1,
                                  c: 2
                                }, Jest.ExpectJs[/* not_ */23](Jest.ExpectJs[/* expect */0]({
                                          a: 1,
                                          b: 2,
                                          c: 3
                                        })));
                    }));
      }));

/*  Not a pure module */
